operations/configuration_management/Write_Ansible_Code_for_CFS.md

```diff
@@ -38,15 +38,74 @@ CFS will handle scaling up Ansible to run on many hosts, but there are still pla
 * See [Target Ansible Tasks for Image Customization](Target_Ansible_Tasks_for_Image_Customization.md) for more information on writing for image customization.
 
 ### Disable fact gathering
 
 * Turn off facts that are not needed in a playbook by setting `gather_facts: false`. If only a few facts are required, it is also possible to limit fact gathering by setting `gather_subset`.
 For more information on `gather_subset`, see the external [Ansible module setup](https://docs.ansible.com/ansible/latest/modules/setup_module.html) documentation.
-* Avoid importing playbooks in other playbooks.  This will trigger fact gathering for each imported playbook, potentially collecting the same information multiple times.
+* Reducing fact gathering time is especially important when importing multiple playbooks from a top level playbook.
+  Fact gathering will trigger for each imported playbook, potentially collecting the same information multiple times.
+  
+### `group_by` and `add_host`
+
+The `group_by` and `add_host` modules can both be used to dynamically generate new hosts groups for the Ansible inventory.
+These modules prove when hosts can be group according to a common property. Then, plays can be designed to only target that particular group.
+Examples of this might include grouping by operating system, hardware type, or a hardware property such as the presence of a GPU.
+
+`group_by` should by used when there are multiple named groups that hosts can be grouped by.
+For more information on using this module see See Ansible's [playbooks best practices](https://docs.ansible.com/ansible/latest/user_guide/playbooks_best_practices.html#handling-os-and-distro-differences) guide.
+
+Example of `group_by`:
+
+```yaml
+- name: group by OS
+  hosts: all
+  tasks:
+    - name: Classify hosts by OS
+      group_by:
+        key: os_{{ ansible_facts['distribution'] }}
+
+- name: centOS playbook
+  hosts: os_CentOS
+  roles:
+    ...
+
+```
+
+`add_host` is useful for cases where the property is true or false.  It allows users to create a new group consisting of only the hosts where the property is true.  See Ansible's documentation on [](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/add_host_module.html) for more information.
+
+Example of `add_host`:
+
+```yaml
+- name: group by a sample variable
+  hosts: all
+  tasks:
+    - name: Add all hosts where sample_var is true to the new Sample group
+      add_host:
+        name: '{{ inventory_hostname }}'
+        groups: sample_group
+      when: sample_var
+
+- name: Sample playbook
+  hosts: sample_group
+  roles:
+    ...
+```
+
+To target only a subset of a set of nodes, plays should use the following syntax.  In this example the play is targeting only _images_ for `Compute` nodes. `&` takes the intersection of the `Compute` and `sample_group` groups.
```

Comment:
Something went awry in the editing. You are using sample_group for the group when I think you want to use cfs_image. You do this in both image customization and node personalization examples. The text description references _images_ and _running_ which doesn't correlate with a random sample_group group.